concrete-core/src/backends/core/private/crypto/ggsw/fourier.rs

```diff
@@ -0,0 +1,725 @@
+use crate::backends::core::private::math::fft::{Complex64, FourierPolynomial};
+use crate::backends::core::private::math::polynomial::Polynomial;
+use crate::backends::core::private::math::tensor::{
+    ck_dim_div, ck_dim_eq, AsMutSlice, AsMutTensor, AsRefSlice, AsRefTensor, IntoTensor, Tensor,
+};
+use crate::backends::core::private::math::torus::UnsignedTorus;
+use crate::backends::core::private::utils::{zip, zip_args};
+
+use super::{GgswLevelMatrix, StandardGgswCiphertext};
+
+use crate::backends::core::private::math::decomposition::{DecompositionLevel, SignedDecomposer};
+use concrete_commons::parameters::{
+    DecompositionBaseLog, DecompositionLevelCount, GlweSize, PolynomialSize,
+};
+use concrete_fftw::array::AlignedVec;
+#[cfg(feature = "multithread")]
+use rayon::{iter::IndexedParallelIterator, prelude::*};
+
+/// A GGSW ciphertext in the Fourier Domain.
+#[derive(Debug, Clone, PartialEq)]
+pub struct FourierGgswCiphertext<Cont, Scalar> {
+    tensor: Tensor<Cont>,
+    poly_size: PolynomialSize,
+    rlwe_size: GlweSize,
+    decomp_base_log: DecompositionBaseLog,
+    _scalar: std::marker::PhantomData<Scalar>,
+}
+
+impl<Scalar> FourierGgswCiphertext<AlignedVec<Complex64>, Scalar> {
+    /// Allocates a new GGSW ciphertext in the Fourier domain whose coefficients are all `value`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use concrete_commons::parameters::{
+    ///     DecompositionBaseLog, DecompositionLevelCount, GlweSize, PolynomialSize,
+    /// };
+    /// use concrete_core::backends::core::private::crypto::ggsw::FourierGgswCiphertext;
+    /// use concrete_core::backends::core::private::math::fft::Complex64;
+    /// let ggsw: FourierGgswCiphertext<_, u32> = FourierGgswCiphertext::allocate(
+    ///     Complex64::new(0., 0.),
+    ///     PolynomialSize(10),
+    ///     GlweSize(7),
+    ///     DecompositionLevelCount(3),
+    ///     DecompositionBaseLog(4),
+    /// );
+    /// assert_eq!(ggsw.glwe_size(), GlweSize(7));
+    /// assert_eq!(ggsw.decomposition_level_count(), DecompositionLevelCount(3));
+    /// assert_eq!(ggsw.decomposition_base_log(), DecompositionBaseLog(4));
+    /// ```
+    pub fn allocate(
+        value: Complex64,
+        poly_size: PolynomialSize,
+        rlwe_size: GlweSize,
+        decomp_level: DecompositionLevelCount,
+        decomp_base_log: DecompositionBaseLog,
+    ) -> Self
+    where
+        Scalar: Copy,
+    {
+        let mut tensor = Tensor::from_container(AlignedVec::new(
+            decomp_level.0 * rlwe_size.0 * rlwe_size.0 * poly_size.0,
+        ));
+        tensor.as_mut_tensor().fill_with_element(value);
+        FourierGgswCiphertext {
+            tensor,
+            poly_size,
+            rlwe_size,
+            decomp_base_log,
+            _scalar: Default::default(),
+        }
+    }
+}
+
+impl<Cont, Scalar> FourierGgswCiphertext<Cont, Scalar> {
+    /// Creates a GGSW ciphertext in the standard domain from an existing container.
```

Comment:
is it not Fourier domain and not standard domain?

Reply:
It is indeed, I'll fix that.
(No, that was not pasted from somewhere else :face_with_rolling_eyes:)